internal/pkg/term/progress/component.go

```diff
@@ -158,3 +163,136 @@ func renderComponents(out io.Writer, components []Renderer) (numLines int, err e
 	}
 	return numLines, nil
 }
+
+// SummaryBarComponent returns a summary bar given data and the string representations of each data category.
+// For example, data[0] will be represented by representations[0] in the summary bar.
+// If len(representations) < len(data), the default representation "□" is used for all data category with missing representation.
+type SummaryBarComponent struct {
+	Length              int      // Length of the summary bar.
+	Data                []int    // Data to draw using the summary bar. The order matters, e.g. the value in position 0 will be the leftmost portion of the bar.
+	Representations     []string // Representations of each data value. Must be at least as long as `Data`. For example, Data[0] will be represented by Representations[0] in the summary bar.
+	EmptyRepresentation string   // Representation to use for an empty bar.
+}
+
+// NewSummaryBarComponent returns a SummaryBarComponent.
+func NewSummaryBarComponent(length int, data []int, representations []string, emptyRepresentation string) (*SummaryBarComponent, error) {
```

Comment:
Should we make the `type SummaryBarComponent` private and instead have a constructor like this?
```suggestion
func SummaryBarComponent(length int, data []int, representations []string, emptyRepresentation string) (Renderer, error) {
```
This is a rare case where it's okay I think to return an interface as all these types actually implement this interface

Reply:
What do you think of moving this struct to package under `progress/summarybar/`?

```go
package summarybar

type Datum struct { // This type ensures that len(rep) == len(value)
   Representation string
   Value int
} 

// Optional config to override defaults.
type Opts struct {
   Width int // Number of characters used by the summary bar. Defaults to 10.
   EmptyRep string // / Representation to use for an empty bar. Defaults to " ".
}

func New(data []Datum) progress.Renderer
func NewWithOpts(data []Datum, opts Opts) progress.Renderer
```

I'm thinking also removing the validation in the constructor and then blowing up in `Render` instead would simplify things for the client

Reply:
Or we can just have one constructor
```go
type data struct {
   representation string
   value int
   width int
   emptyRep string
}
type Opt func(data)
func WithWidth(width int) Opt {
  return func(d data) {
    d.width = width
  }
}
func New(data []Datum, opts ...Opt) progress.Renderer
```